reconcile/test/test_status_page_components.py

```diff
@@ -67,296 +69,295 @@ def f(self):
+    """
+    Test if state management correctly relates component ids with
+    components declared in desired state.
+    """
+    fixture_name = "test_state_management_on_fetch.yaml"
+    apply_mock = mocker.patch.object(AtlassianStatusPage, 'apply_component',
+                                     autospec=True)
+    apply_mock.return_value = None
+    fetch_mock = mocker.patch.object(AtlassianStatusPage, '_fetch_components',
+                                     autospec=True)
+    fetch_mock.return_value = []
+    mocker.patch.object(VaultSecretRef, '_resolve_secret',
+                        new_callable=stub_resolve_secret)
+
+    page = get_page_fixtures(fixture_name)[0]
+    page.reconcile(True, get_state_fixture(fixture_name))
+
+    for c in page.components:
+        if c.name == "comp_1":
+            assert c.component_id == "comp_id_1"
+        elif c.name == "comp_2":
+            assert c.component_id is None
+
+
+def test_state_management_on_bind(mocker):
+    """
+    Test if state management correctly binds components with their
+    corresponding id on the status page.
+    """
+    fixture_name = "test_state_management_on_bind.yaml"
+    fetch_mock = mocker.patch.object(AtlassianStatusPage, '_fetch_components',
+                                     autospec=True)
+    fetch_mock.return_value = \
+        get_atlassian_component_fixtures(fixture_name, "page_1")
+    mocker.patch.object(VaultSecretRef, '_resolve_secret',
+                        new_callable=stub_resolve_secret)
+    state = StateStub()
+
+    page = get_page_fixtures(fixture_name)[0]
+    page.reconcile(False, state)
+
+    assert page.components[0].component_id == "comp_id_1"
+    assert state.get_all("")["comp_1"] == "comp_id_1"
+
+
+def test_dry_run_on_create(mocker):
+    _exec_dry_run_test_on_create(mocker, True)
+
+
+def test_no_dry_run_on_create(mocker):
+    _exec_dry_run_test_on_create(mocker, False)
+
+
+def _exec_dry_run_test_on_create(mocker, dry_run):
+    create_mock = mocker.patch.object(AtlassianStatusPage, "_create_component",
+                                      autospec=True)
+    create_mock.return_value = "id"
+    fetch_mock = mocker.patch.object(AtlassianStatusPage, "_fetch_components",
+                                     autospec=True)
+    fetch_mock.return_value = []
+    provider = AtlassianStatusPage(page_id="page_id",
+                                   api_url="https://a.com",
+                                   token="token")
 
-        with self.assertRaises(ValueError):
-            page.update_component_status(True, "comp_1", "invalid", state)
+    component = StatusComponent(name="comp_1", displayName="comp_1",
+                                groupName=None)
+    provider.apply_component(dry_run, component)
+
+    if dry_run:
+        create_mock.assert_not_called()
+    else:
+        create_mock.assert_called()
+
+
+def test_dry_run_on_update(mocker):
+    _exec_dry_run_test_on_update(mocker, True)
```

Comment:
This isn't really specific to moving to pytest and therefore not a blocker, but I think tests like this are a bit obscure. It's not really clear what's going on here until I go off and read `_exec_dry_run_test_on_update`. I'm not sure whether that level of DRY is really helpful or not.

Reply:
i know what you mean. at the same time i always have this nagging feeling at the back of my head what bad things can happen when we have a bug somewhere not propagating dry_run information properly.